File: flyweight/typescript/blue-box.ts
import {Box} from "./box";
import {BoxColor} from "./box-color";

// Flyweight (Box) şablonunu uygulayan sınıftır.
// UML diyagramındaki ConcreteFlyweight sınıfına denk gelmektedir.
export class BlueBox extends Box {
  constructor(readonly width, readonly height) {
    super();
    this.width = width;
    this.height = height;
    this.boxColor = BoxColor.BLUE;
  }

  public draw(locationX: number, locationY: number): void {
    console.log(`${this.boxColor} box drawn. ${locationX},${locationY}`)
  }
}

File: flyweight/go/blue-box.go
package main

import "fmt"

// Flyweight (Box) şablonunu uygulayan sınıftır.
// UML diyagramındaki ConcreteFlyweight sınıfına denk gelmektedir.
type blueBox struct {
	width    int
	height   int
	boxColor int
}

func NewBlueBox(width int, height int) *blueBox {
	return &blueBox{
		width:    width,
		height:   height,
		boxColor: BLUE,
	}
}

func (b blueBox) Draw(locationX, locationY int) {
	fmt.Printf("Blue box drawn. %d,%d\n", locationX, locationY)
}
